### PR TITLE
NEW Add support for callable link formatting

### PR DESCRIPTION
**Pull request by wilr**

### CODE REVIEW

File: code/Report.php [hunk 1]
```diff
@@ -97,6 +99,13 @@ class Report extends ViewableData
         SideReportWrapper::class,
     ];
 
+    public function __construct()
+    {
+        $this->params = isset($_REQUEST['filters']) ? $_REQUEST['filters'] : array();
```
Comment (tractorcow) [inline code/Report.php]: This change doesn't actually implement the given TODO

```php
// TODO Remove coupling with global state
```

This in fact makes this code riskier; It relies on the global state a lot earlier (object construction, not at render time). This makes testing more susceptible to global state change.

File: code/Report.php [hunk 2]
```diff
@@ -324,17 +338,14 @@ public function getCMSActions()
      */
     public function getReportField()
     {
-        // TODO Remove coupling with global state
-        $params = isset($_REQUEST['filters']) ? $_REQUEST['filters'] : array();
-        $items = $this->sourceRecords($params, null, null);
```
Comment (tractorcow) [inline code/Report.php]: We could use `Injector::inst()->get(HTTPRequest::class)` instead to get the request.

File: code/Report.php (hunk 1 above)
Comment (wilr) [reply code/Report.php]: @tractorcow The goal of this commit wasn't to implement the TODO and I can put it back.

Comment (tractorcow) [reply code/Report.php]: This still leaves this concern;

> This in fact makes this code riskier; It relies on the global state a lot earlier (object construction, not at render time). This makes testing more susceptible to global state change.

**Comment by robbieaverill** [comment]
Quick note: we've implemented this functionality using the build in `formatting` callbacks instead in the elemental-blocks module, so this PR no longer blocks that. Having a callable link argument would still be nice though going forward.

**Comment by tractorcow** [comment]
Do you still need this PR @wilr ?

**Comment by wilr** [comment]
@tractorcow as per Robbie's message - be nice having it going forward.

**Comment by tractorcow** [comment]
In that case can you please implement the feedback @wilr and I'll re-review.

File: code/Report.php [hunk 3]
```diff
@@ -324,17 +341,14 @@ public function getCMSActions()
...
+        $items = $this->sourceRecords($this->params, null, null);
 
         $gridFieldConfig = GridFieldConfig::create()->addComponents(
 
             new GridFieldButtonRow('before'),
             new GridFieldPrintButton('buttons-before-left'),
             new GridFieldExportButton('buttons-before-left'),
             new GridFieldToolbarHeader(),
-            new GridFieldSortableHeader(),
```
Comment (flamerohr) [inline code/Report.php]: how come `GridFieldSortableHeader` was removed?

**Comment by flamerohr** [comment]
@wilr I've refactored the changes a bit so it lazily calls for request, let me know if this would work for you

**Comment by flamerohr** [comment]
@tractorcow would you mind giving it a once over review?

**Comment by tractorcow** [comment]
image

**Comment by robbieaverill** [comment]
Updated

**Comment by tractorcow** [comment]
Ok I had some other fixes to clean up this class, slightly.

This class has some terribly rough code in it. XD

**Pull request merged**